### PR TITLE
Sites history table

### LINKED ISSUE #165: Site history table
## Detailed Description
It would be good to store the history of site changes. 

## Context
- good to know who changes the sites

## Possible Implementation
- create site history table
- same fields as site 
- extra field: user_uuid, the user who updated it
- extra field: update_from_site_historic_uuid, a link to the last update of site parameters
- link this table to the sites table
- have function that 1. updates site, 2. add new enter to site history table.

### PR DESCRIPTION
**Pull request by mduffin95**
# Pull Request

## Description

Introduces a history table for sites. The table is updated by a trigger that calls the `log_site_changes` function. The function stores the new record (in the case of INSERT and UPDATE) or the old record (in the case of DELETE) in binary json format (JSONB). 

Fixes #165 

## How Has This Been Tested?

In #169 I ensured that the tests use the alembic migrations. This means the function and trigger are present during tests. I've then extended the tests to check that the history table has been updated.

## Checklist:

- [x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by mduffin95** [comment]
work in progress ⚒️

File: alembic/versions/12c0ccf6b825_create_trigger.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = '12c0ccf6b825'
+down_revision = 'c8ef88c250e7'
+branch_labels = None
+depends_on = None
+
+create_trigger = """CREATE OR REPLACE FUNCTION log_site_changes()
```
Comment (peterdudfield) [inline alembic/versions/12c0ccf6b825_create_trigger.py:18]: This is very nice, and i really like that all the logic is SQL, thanks so much for doing this.

File: alembic/versions/12c0ccf6b825_create_trigger.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        ) VALUES (
+            gen_random_uuid(),
+            NEW.site_uuid,
+            to_jsonb(NEW),
+            user_uuid,
+            'INSERT',
+            NOW()
+        );
+
+    ELSIF TG_OP = 'UPDATE' THEN
```
Comment (peterdudfield) [inline alembic/versions/12c0ccf6b825_create_trigger.py:42]: coudl this first two cases come together, and the `operation_type` be `TG_OP`. I have to admit this is not my area of expertise though

File: tests/test_delete.py [hunk 3]
```diff
@@ -17,6 +17,12 @@ def test_delete_site(db_session):
 
     site_1 = make_fake_site(db_session=db_session, ml_id=1)
 
+    # history table should contain a single entry
```
Comment (peterdudfield) [inline tests/test_delete.py:20]: nice

File: tests/test_delete.py [hunk 4]
```diff
@@ -31,6 +37,11 @@ def test_delete_site(db_session):
     assert site is None
     assert message == f"Site with site uuid {site_uuid} deleted successfully"
 
+    deleted_site = (
```
Comment (peterdudfield) [inline tests/test_delete.py:40]: nice

File: tests/write/test_write.py [hunk 5]
```diff
@@ -295,21 +342,40 @@ def test_save_api_call_to_db(db_session):
...
         site_uuid=str(site.site_uuid),
         site_info=metadata_to_update,
+        user_uuid=user.user_uuid,
     )
 
     assert site.tilt == metadata_to_update.tilt
     assert site.capacity_kw == metadata_to_update.capacity_kw
     assert site.latitude == prev_latitude
 
+    # after editing there should be another entry in the history table
```
Comment (peterdudfield) [inline tests/write/test_write.py:371]: is it worth also checking that tehre are two entries in the table?

File: pvsite_datamodel/sqlmodels.py [hunk 6]
```diff
@@ -209,6 +209,33 @@ class SiteSQL(Base, CreatedMixin):
...
 
 
+class SiteHistorySQL(Base, CreatedMixin):
+    """Class representing the sites history table.
+
+    Stores a history of changes to sites over time. Uses JSONB so that schema changes to the
+    site table do not affect the history table.
+    """
+
+    __tablename__ = "h_sites"
```
Comment (peterdudfield) [inline pvsite_datamodel/sqlmodels.py:219]: shoule this be slightly more verbose, maybe `sites_history`

**Review by peterdudfield — CHANGES_REQUESTED** [review]
Thank you so much @mduffin95 for this. This is really nice and a neat way to do it. 

There are some very small comments in the PR. You ok to address them? I'm happy to take them on if you dont want to

**Comment by mduffin95** [comment]
Yeah happy to update it 👍

File: pvsite_datamodel/sqlmodels.py (hunk 6 above)
Comment (mduffin95) [reply pvsite_datamodel/sqlmodels.py:219]: Yeah I looked around and went with `h_` as it seems to be a common convention but happy to update it as you're right `_history` is a bit clearer

File: alembic/versions/12c0ccf6b825_create_trigger.py (hunk 1 above)
Comment (mduffin95) [reply alembic/versions/12c0ccf6b825_create_trigger.py:18]: No problem! Glad you like it.

**Review by peterdudfield — APPROVED** [review]

**Comment by peterdudfield** [comment]
Thank you very much @mduffin95

**Review by devsjc — APPROVED** [review]
Thanks @mduffin95, this looks great!

**Pull request merged**